pypiserver/backend.py

```diff
@@ -0,0 +1,224 @@
+            lambda pkg: pkg.pkgname == name and pkg.version == version,
+            self.get_all_packages(),
+        )
+
+
+class SimpleFileBackend(Backend):
+    def __init__(self, config: Configuration, roots: t.List[PathLike]):
+        super().__init__(config)
+        self.roots = [Path(root).resolve() for root in roots]
+
+    def get_all_packages(self):
+        return itertools.chain.from_iterable(listdir(r) for r in self.roots)
+
+    def add_package(self, filename: str, fh: t.BinaryIO):
+        as_file(fh, self.roots[0].joinpath(filename))
+
+    def remove_package(self, pkg: PkgFile):
+        os.remove(pkg.fn)
+
+    def exists(self, filename):
+        # TODO: Also look in subdirectories?
+        return any(root.joinpath(filename).exists() for root in self.roots)
+
+
+class CachingFileBackend(SimpleFileBackend):
+    def __init__(
+        self, config: Configuration, roots: t.List[PathLike], cache_manager
+    ):
+        super().__init__(config, roots)
+
+        self.cache_manager: CacheManager = cache_manager
+
+    def get_all_packages(self):
+        return itertools.chain.from_iterable(
+            self.cache_manager.listdir(r, listdir) for r in self.roots
+        )
+
+    def digest(self, pkg: PkgFile):
+        return self.cache_manager.digest_file(
+            pkg.fn, self.hash_algo, digest_file
+        )
+
+
+def as_file(fh: t.BinaryIO, destination: PathLike):
+    """write a byte stream into a destination file. Writes are chunked to reduce
+    the memory footprint
+    """
+    chunk_size = 2 ** 20  # 1 MB
+    offset = fh.tell()
+    try:
+        with open(destination, "wb") as dest:
+            for chunk in iter(lambda: fh.read(chunk_size), b""):
+                dest.write(chunk)
+    finally:
+        fh.seek(offset)
+
+
+def listdir(root: PathLike) -> t.Iterable[PkgFile]:
+    root = Path(root).resolve()
+    for dirpath, dirnames, filenames in os.walk(root):
+        dirnames[:] = [x for x in dirnames if is_allowed_path(x)]
+        for x in filenames:
+            fn = os.path.join(root, dirpath, x)
+            if not is_allowed_path(x) or not Path(fn).is_file():
+                continue
+            res = guess_pkgname_and_version(x)
+            if not res:
+                # Seems the current file isn't a proper package
+                continue
+            pkgname, version = res
+            if pkgname:
+                yield PkgFile(
+                    pkgname=pkgname,
+                    version=version,
+                    fn=fn,
+                    root=root,
+                    relfn=fn[len(str(root)) + 1 :],
+                )
+
+
+def digest_file(fpath, hash_algo: str):
+    """
+    Reads and digests a file according to specified hashing-algorith.
+
+    :param hash_algo: any algo contained in :mod:`hashlib`
+    :return: <hash_algo>=<hex_digest>
+
+    From http://stackoverflow.com/a/21565932/548792
+    """
+    blocksize = 2 ** 16
+    digester = getattr(hashlib, hash_algo)()
```

Comment:
let's pull in my update from #349, which essentially just replace this line with
```suggestion
    digester = hashlib.new(hash_algo)
```
to support a wider variety of hashing algorithms